### ISSUE TITLE
(RuntimeError) Wallaby had an internal issue with HTTPoison

### ISSUE THREAD

**Issue by stephaniewilkinson** [elixir-wallaby/wallaby#365]
## Issue

When I switch to using Chrome, I get this error:

```
     ** (RuntimeError) Wallaby had an internal issue with HTTPoison
     stacktrace:
       (wallaby) lib/wallaby/httpclient.ex:32: Wallaby.HTTPClient.make_request/4
       (wallaby) lib/wallaby/experimental/chrome.ex:64: Wallaby.Experimental.Chrome.start_session/1
       (wallaby) lib/wallaby.ex:80: Wallaby.start_session/1
       (school) test/support/feature_case.ex:25: SchoolWeb.FeatureCase.__ex_unit_setup_0/1
       (school) test/support/feature_case.ex:1: SchoolWeb.FeatureCase.__ex_unit__/2
       test/school_web/features/view_test.exs:1: SchoolWeb.UserListTest.__ex_unit__/2
```

Versions:
`elixir: "~> 1.4",
`

## Test Code & HTML

Here's my test:

```
defmodule SchoolWeb.UserListTest do
  use SchoolWeb.FeatureCase, async: true

  import Wallaby.Query, only: [css: 2]

  test "homepage is accessible", %{session: session} do
    session
    |> visit("/")
    |> assert_has(css(".header", count: 1, text: "Welcome"))
  end
end
```

Here's my FeatureCase:
```
defmodule SchoolWeb.FeatureCase do
  use ExUnit.CaseTemplate

  using do
    quote do
      use Wallaby.DSL

      alias School.Repo
      import Ecto
      import Ecto.Changeset
      import Ecto.Query

      import SchoolWeb.Router.Helpers
    end
  end

  setup tags do
    :ok = Ecto.Adapters.SQL.Sandbox.checkout(School.Repo)

    unless tags[:async] do
      Ecto.Adapters.SQL.Sandbox.mode(School.Repo, {:shared, self()})
    end

    metadata = Phoenix.Ecto.SQL.Sandbox.metadata_for(School.Repo, self())
    {:ok, session} = Wallaby.start_session(metadata: metadata)
    {:ok, session: session}
  end
end
```

Here's my HTML:

`  <h1 class="header center">Welcome to School!</h1>
`
Thanks for any help you can give!

**Comment by keathley** [comment]
Thanks for the report @stephaniewilkinson. Can you confirm what version of chrome and chromedriver you have installed?

**Comment by stephaniewilkinson** [comment]
My Chrome is Version 66.0.3359.117 (Official Build) (64-bit)
My ChromeDriver is 2.37.544337 (8c0344a12e552148c185f7d5117db1f28d6c9e85)
I had previously tried on Chrome 65 with the same result.

For phantomjs I didn't customize the version, just ran this:

`npm install -g phantomjs-prebuilt
`

**Comment by keathley** [comment]
Thanks! All of that looks correct to me. We get those httppoison errors when there’s an issue communicating with chromedriver. Based on the stack trace it looks like it’s not able to create a new session which typically indicates an issue with wallabys ability to talk to chromedriver or chromedrivers ability to talk to chrome.

**Comment by stephaniewilkinson** [comment]
Thanks! any ideas what to try next?

**Comment by keathley** [comment]
What OS are you using? Are both chrome and chromedriver in your path? Other things to try might be trying out the latest elixir and otp version.

**Comment by dsdshcym** [comment]
Got an similar error:

```
     ** (RuntimeError) Wallaby had an internal issue with HTTPoison
     code: |> assert_has(css("td", text: node_id_1))
     stacktrace:
       (wallaby) lib/wallaby/httpclient.ex:32: Wallaby.HTTPClient.make_request/4
       (wallaby) lib/wallaby/experimental/selenium/webdriver_client.ex:125: Wallaby.Experimental.Selenium.WebdriverClient.text/1
       (wallaby) lib/wallaby/driver/log_checker.ex:6: Wallaby.Driver.LogChecker.check_logs!/2
       (wallaby) lib/wallaby/browser.ex:910: Wallaby.Browser.matching_text?/2
       (elixir) lib/enum.ex:2872: Enum.filter_list/2
       (elixir) lib/enum.ex:2873: Enum.filter_list/2
       (wallaby) lib/wallaby/browser.ex:903: Wallaby.Browser.validate_text/2
       (wallaby) lib/wallaby/browser.ex:925: anonymous fn/3 in Wallaby.Browser.execute_query/2
       (wallaby) lib/wallaby/br
                       _build/test/lib/wallaby/priv/run_command.sh: line 7: 85966 Segmentation fault: 11  "$@"
```

Not sure if it's the same issue though.

I'm running macOS 10.13.5, Chrome 67.0.3396.79, ChromeDriver 2.40.565386, Elixir 1.6.5 (compiled with OTP 20).

**Comment by keathley** [comment]
We need to add more backoff and jitter to our http client. We see these issues when chromedriver starts to get overwhelmed with requests.

**Comment by calvinkw1** [comment]
@keathley any update here? Am running into the same issue:

```** (RuntimeError) Wallaby had an internal issue with HTTPoison
     stacktrace:
       (wallaby) lib/wallaby/httpclient.ex:32: Wallaby.HTTPClient.make_request/4
       (wallaby) lib/wallaby/experimental/chrome.ex:64: Wallaby.Experimental.Chrome.start_session/1
       (wallaby) lib/wallaby.ex:80: Wallaby.start_session/1
       (orderbook) test/support/browser_feature_case.ex:17: Test.Support.Browser.Feature.Case.__ex_unit_setup_all_0/1
       (orderbook) test/support/browser_feature_case.ex:1: Test.Support.Browser.Feature.Case.__ex_unit__/2
       test/orderbook_web/features/browser/deposit_address_test.exs:1: OrderbookWeb.Features.Browser.DepositAddressTests.__ex_unit__/2```

**Comment by calvinkw1** [comment]
@keathley any update here? Am running into the same issue:

```
** (RuntimeError) Wallaby had an internal issue with HTTPoison
     stacktrace:
       (wallaby) lib/wallaby/httpclient.ex:32: Wallaby.HTTPClient.make_request/4
       (wallaby) lib/wallaby/experimental/chrome.ex:64: Wallaby.Experimental.Chrome.start_session/1
       (wallaby) lib/wallaby.ex:80: Wallaby.start_session/1
       (orderbook) test/support/browser_feature_case.ex:17: Test.Support.Browser.Feature.Case.__ex_unit_setup_all_0/1
       (orderbook) test/support/browser_feature_case.ex:1: Test.Support.Browser.Feature.Case.__ex_unit__/2
       test/redacted/features/browser/redacted.exs:1: Redacted.Features.Browser.Redacted.__ex_unit__/2

```
I have chromedriver 2.36.540469 installed, Elixir 1.6.4 (compiled with OTP 20), Mac OS 10.13.5

**Comment by keathley** [comment]
I'm working on tracking down the root cause but so far no luck. Without a way to reliably re-produce this issue locally its very difficult to ensure we have a proper fix. If you have a test case that you can provide that would greatly assist the effort here.

**Comment by keathley** [comment]
I've opened a PR that _may_ help to reduce some of the issues that folks are seeing. If you would like to test out that branch locally that would be very useful.

**Comment by keathley** [comment]
The PR is now on master so if you'd like to test with that it would be very appreciated.

**Comment by calvinkw1** [comment]
Sorry.. didn't work. Getting a similar error:

```
 1) Redacted.Features.Browser.DepositAddressTests: failure on setup_all callback, test invalidated
     ** (RuntimeError) Wallaby had an internal issue with HTTPoison:
     %HTTPoison.Error{id: nil, reason: :econnrefused}
     stacktrace:
       (wallaby) lib/wallaby/httpclient.ex:37: Wallaby.HTTPClient.make_request/5
       (wallaby) lib/wallaby/experimental/chrome.ex:72: Wallaby.Experimental.Chrome.start_session/1
       (wallaby) lib/wallaby.ex:80: Wallaby.start_session/1
       ...
```

This seems to be only for me.. works fine on Circle CI and on other colleagues' machines. Not sure why this is.

I can provide whatever you need but note that I'm an Elixir newb...

**Comment by calvinkw1** [comment]
Sorry.. didn't work. Getting a similar error:

```
 1) Redacted.Features.Browser.AddressTests: failure on setup_all callback, test invalidated
     ** (RuntimeError) Wallaby had an internal issue with HTTPoison:
     %HTTPoison.Error{id: nil, reason: :econnrefused}
     stacktrace:
       (wallaby) lib/wallaby/httpclient.ex:37: Wallaby.HTTPClient.make_request/5
       (wallaby) lib/wallaby/experimental/chrome.ex:72: Wallaby.Experimental.Chrome.start_session/1
       (wallaby) lib/wallaby.ex:80: Wallaby.start_session/1
       ...
```

This seems to be only for me.. works fine on Circle CI and on other colleagues' machines. Not sure why this is.

I can provide whatever you need but note that I'm an Elixir newb...

**Comment by evax** [comment]
I've been having a similar issue and realized that a lot of chrome and chromedriver processes had been accumulating - killing those fixes it for me.

**Comment by keathley** [comment]
@evax Like they're being orphaned after running tests?

@calvinkw1 Can you provide any other insights into what might be different on your machine? Different OS, chromedriver version, chrome version, etc.? Based on the error you posted it looks like chromedriver isn't even starting, hence the `:econnrefused`. Are you sure that chromedriver is in your path? If we can narrow down why thats happening then we might be able to provide a more meaningful error.

**Comment by evax** [comment]
@keathley yes, though in that case if that helps I wasn't running the full test suite but giving `mix test` the path of the test file as an argument, in an umbrella.

**Comment by keathley** [comment]
I wonder if the cleanup script isn't being run correctly because of the umbrella or something similar.

**Comment by evax** [comment]
It could be because the path is relative to the app, so when you have multiple apps in an umbrella, it will likely only work for one app, the others display: `Test patterns did not match any file: path/to/testfile_test.exs`

**Comment by calvinkw1** [comment]
Ah shoot. This is probably something I should have mentioned as well. I was passing in a specific file path rather than running the entire test suite. Running the entire test suite passes without issue.

```
Mac OS High Sierra 10.13.5

Google Chrome is up to date
Version 67.0.3396.99 (Official Build) (64-bit)

chromedriver --version
ChromeDriver 2.36.540469 (1881fd7f8641508feb5166b7cae561d87723cfa8)
```

Watching `htop`, I don't see a `chromedriver` process run when I only pass in a file path, but I do when I run the full suite.

Also, if it matters, I work on a Ruby project that has `ChromeDriver 2.40.565386 (45a059dc425e08165f9a10324bd1380cc13ca363)` installed.

**Comment by keathley** [comment]
@calvinkw1 The only thing I can think in that case would be that its not running the test_helper.exs when you run that specific file. Or that file isn't using the correct `FeatureCase` or something similar.

**Comment by mjankowski** [comment]
After using chromedriver+wallaby successfully on an elixir test suite for a while, I saw this issue for first time today. I remembered that last week while working on other projects I had fiddled around with my chromedriver install - I had a few ruby projects and wound up using the `chromedriver-helper` gem in them to get local and CI working well together.

Anyway - I have the same chromedriver version installed (`ChromeDriver 2.42.591059 (a3d9684d10d61aa0c45f6723b327283be1ebaad8)`) via both homebrew and as an `rbenv` shim from the gem installs.

When I manipulate my path to get the chromedriver installed via homebrew as the chromedriver that wallaby uses, everything is fine. When I update path to use the rbenv shim, I see the "Wallaby had an internal issue with HTTPoison" sort of errors described here.

I'm not sure what if anything to do next to help debug what these installs are doing differently - but maybe this data point helps someone. Happy to do more if I can be useful.

**Comment by sparta-developers** [comment]
We also see 

```
** (RuntimeError) Wallaby had an internal issue with HTTPoison:
%HTTPoison.Error{id: nil, reason: :econnrefused}
```

when running Wallaby tests in our two apps. In one app, we can set the test concurrency to 2 and the error goes away. In the other app, we always get the error unless we slow things down by increasing the jitter by [forking Wallaby](https://github.com/sparta-science/wallaby/blob/longer-retry-jitter/lib/wallaby/httpclient.ex#L10)  and by passing in our own implementation of `create_session` that sleeps for 500ms inside a [mutex](https://github.com/niahoo/mutex):

```elixir
def create_session(base_url, capabilities) do
  Mutex.under(:chrome_session_mutex, :start_session, :infinity, fn ->
    :timer.sleep(500)

    Wallaby.Experimental.Selenium.WebdriverClient.create_session(base_url, capabilities)
  end)
end
```

We're using chromedriver 74.0.3729.6 installed via homebrew on MacOS. 

Are other people having success with chromedriver?

**Comment by keathley** [comment]
I suspect that chromedriver just can't keep up with the amount of traffic we're sending it at the same time. We could try to force this to behave better by setting limits on the hackney pool so that we don't overwhelm chromedriver as often.

**Comment by keathley** [comment]
I haven't tried this but I wonder if the new puppet api thing that google has would allow for more performance and be more reliable then the current chromedriver setup.

**Comment by sparta-developers** [comment]
> We could try to force this to behave better by setting limits on the hackney pool so that we don't overwhelm chromedriver as often.

@keathley we tried setting hackney's pool size to 1, which unfortunately did not help. (It's possible that we're doing it wrong, but we checked a few different ways that it was using our pool size.) 

We have learned that we're getting the "connection refused" messages because chromedriver crashes but we haven't yet found anything useful in its logs.

> I haven't tried this but I wonder if the new puppet api thing that google has would allow for more performance and be more reliable then the current chromedriver setup.

Do you mean [Puppeteer](https://github.com/GoogleChrome/puppeteer)? We see that there are a few issues (#417, #358) about that already. One knock against it was that it's Chrome-only, but maybe it's worth doing anyway?

**Comment by keathley** [comment]
I'd be in favor of moving to puppeteer for chrome only and using webdriver for selenium. We don't have a ton of issues with selenium handling the webdriver stuff.

**Comment by AndroidOatmeal** [comment]
Any updates on this issue? This problem unfortunately causes my test suite to fail about 50% of the time right now.

**Comment by michallepicki** [comment]
@AndroidOatmeal Could you provide more background, or are your stacktraces exactly the same as in the issue description? Does the provided example reproduce the issue with similar rate, or maybe you could provide a different minimal example? What are your Elixir, Chrome and Chromedriver versions?

**Comment by AndroidOatmeal** [comment]
Sure, here are my versions

```
Elixir 1.8.1 (compiled with Erlang/OTP 20)
ChromeDriver 74.0.3729.6 (255758eccf3d244491b8a1317aa76e1ce10d57e9-refs/branch-heads/3729@{#29})
Chrome Version 75.0.3770.100 (Official Build) (64-bit)```

My errors are the same as described above. They seem to happen intermittently. Our acceptance test suite has 47 tests currently.

```
     ** (RuntimeError) Wallaby had an internal issue with HTTPoison:
     %HTTPoison.Error{id: nil, reason: :econnrefused}
     stacktrace:
       (wallaby) lib/wallaby/httpclient.ex:37: Wallaby.HTTPClient.make_request/5
       (wallaby) lib/wallaby/experimental/chrome.ex:79: Wallaby.Experimental.Chrome.start_session/1
       (wallaby) lib/wallaby.ex:80: Wallaby.start_session/1```

**Comment by AndroidOatmeal** [comment]
@keathley @michallepicki @sparta-developers  Do you have any suggestions for debugging this issue? I'm not afraid to get my hands dirty to figure this out, I just don't know where in the Wallaby codebase to begin.

**Comment by kerryb** [comment]
I'm also seeing this error. It only happens when I run tests with `async: true` (I'm trying to migrate away from Cabbage, which only worked with `async: false`). Even with just two tests running concurrently, I get the error almost every time.

Also happy to help debug, if you point me in the right direction.

**Comment by keathley** [comment]
I'd start by trying to trace the calls to HTTPClient and the Wallaby.Experimental.Chrome.Chromedriver process. That's the process that controls chromedriver through a port. My intuition is that we're causing chromedriver to crash or something similar.

@kerryb I think the reason that you're seeing issues when running in async: true mode is either 1) we're overwhelming chromedriver with the number of requests that we're sending it or 2) inducing some internal bug with chromedriver because it can't handle the number of independent sessions we're using. In either case we're overwhelming chromedriver but my money is on 1 being the root cause.

An easy way to test this is to limit the number of http requests we allow through at any given time. Because we're using hackney (through httpoison) a quick way to test this out would be to update the HTTPClient api to use a dedicated hackney pool and set the pool limits arbitrarily low like so:

```
# Add this to out supervision tree
:hackney_pool.child_spec(:wallaby_pool, [timeout: 15000, max_connections: 4])

# Update http client to use hackney pool on line 41 of http_client.ex.
HTTPoison.request(url, body, headers(), hackney: [pool: :wallaby_pool])
```

That's where I would start debugging.

**Comment by mhanberg** [comment]
@keathley would it be beneficial to actually surface HTTPoison errors to the user?

Why did you decide to swallow them in a wallaby error?

**Comment by keathley** [comment]
@mhanberg Yeah we should totally be surfacing them. I'm not sure I had a reason for not surfacing them. Either way its something we should change now.

**Comment by eahanson** [comment]
@mhanberg the Wallaby code _does_ try to surface the HTTPoison errors, and it works in my experience: 

https://github.com/keathley/wallaby/blob/6333301eb41dd63c03cf5ac2370087808c41cd8c/test/wallaby/http_client_test.exs#L75

When I experience this problem (which is always), it's because HTTPoison gets `:econnrefused` which (again in my experience) is because the chromedriver process has crashed. It crashes pretty much any time I send it a lot of requests, like when I'm running tests in parallel.

**Comment by keathley** [comment]
@eahanson Yup that tracks with what I've seen in the past. I think we can try to limit the amount of in flight requests to chromedriver which should help it stay healthy. We might need to work out a better supervision strategy for our process that manages chromedrivers as well. We might have to fail a test case if a chromedriver crashes in the middle of it but we shouldn't need to fail the entire suite.

**Comment by mhanberg** [comment]
@eahanson I am seeing that support for surfacing the error was included a few months after this issue was opened.


@keathley @michallepicki I wonder if we should close this issue in favor of an issue regarding overwhelming chromedriver. What do you think?

**Comment by mhanberg** [comment]
Also, to address the references to Puppeteer above:

Puppeteer is a Node.js library primarily made to write web scraping scripts.

**Comment by michallepicki** [comment]
@mhanberg I think it would be possible to write a Wallaby driver using the Chrome Devtools Protocol through puppeteer... For example there's [taiko](https://taiko.gauge.org/) which is a testing tool that uses it directly.

**Comment by keathley** [comment]
Like @michallepicki said, I think Wallaby could support the chrome devtools protocol for its chrome support. I think that's a good future direction personally but let's set that aside for a moment and see if we can fix the current issue. For now lets try to limit the number of requests hitting chromedriver simultaneously and see if that helps keep chromedriver healthy. If that solves the immediate problem we can close this ticket and focus on moving to the chrome devtools protocol as a more future proof step.

**Comment by mhanberg** [comment]
@michallepicki My point being we wouldn't use Puppeteer, we would write a driver for the Chrome DevTools protocol. I am being pedantic but I want to clarify what I was saying.

**Comment by mhanberg** [comment]
FYI last night I started working on the supervision and http request strategy for chromedriver.

**Comment by wood-archer** [comment]
@mhanberg 
Was this fixed in the last release? (https://github.com/keathley/wallaby/commit/231230585fefece65be1ca4f7760a6c893543810)

**Comment by mhanberg** [comment]
@wood-archer Not explicitly, but it might occur less frequently now. 

Do you mind upgrading and testing it out?

**Comment by wood-archer** [comment]
Sure. will give it a try. Thanks!

**Comment by wood-archer** [comment]
@mhanberg sorry for the long time to post back but it looks like the error `Wallaby had an internal issue with HTTPoison` is gone after updating from wallaby 0.20.0 to 0.23.0. Now I can only see the errors like `** (Wallaby.QueryError) Expected to find 1, visible element that matched the css '# ..` but most likely these are just on my end.

**Comment by AndroidOatmeal** [comment]
For reference, I am still experiencing this issue often, even on `0.23.0`. :(

> FYI last night I started working on the supervision and http request strategy for chromedriver.
@mhanberg Any luck with on this? Do you need any help with it?

**Comment by mhanberg** [comment]
@AndroidOatmeal I think I had some luck, but I (obviously) got distracted from this for some reason.

I'll push a branch tonight and then you can try out that branch.

Sorry about that!

**Comment by mhanberg** [comment]
I have opened a PR that will hopefully help address this. #463 

@AndroidOatmeal would you mind trying out that branch? Thanks!

**Comment by mhanberg** [comment]
I merged #463. I will be able to cut a release tonight. 

If you have run into this problem, please update and let me know if it fixes it for you!

**Comment by AndroidOatmeal** [comment]
@mhanberg Thanks so much for putting time into this. My coworker and I tried the branch, but neither of us saw any noticeable change in the frequency of the errors on our respective machines. Bummer :(

I did an experiment where I ran our suite of 64 acceptance tests 100 times. The suite failed due to this error 43% of the time. When the suite failed, this error usually occurred on all 64 tests at once, though sometimes half of the tests would succeed and the other half would fail with the error.

The error in question:
```
     ** (RuntimeError) Wallaby had an internal issue with HTTPoison:
     %HTTPoison.Error{id: nil, reason: :econnrefused}
     stacktrace:
       (wallaby) lib/wallaby/httpclient.ex:37: Wallaby.HTTPClient.make_request/5
       (wallaby) lib/wallaby/experimental/chrome.ex:96: Wallaby.Experimental.Chrome.start_session/1
       (wallaby) lib/wallaby.ex:79: Wallaby.start_session/1
       (greenlight) test/support/acceptance_test_helpers.ex:74: Greenlight.Web.AcceptanceTestHelpers.build_session_with_host/2
       (greenlight) test/support/acceptance_test_helpers.ex:30: Greenlight.Web.AcceptanceTestHelpers.admin_login/1
```

I'm happy to provide any other information on this issue. Thanks again for your time, it's much appreciated.

**Comment by mhanberg** [comment]
@AndroidOatmeal is this happening on CI as well or only locally?

What Chrome and ChromeDriver versions are you using?

**Comment by AndroidOatmeal** [comment]
@mhanberg Yes, this issue occurs on CircleCI as well.

```
Version 76.0.3809.132 (Official Build) (64-bit)
ChromeDriver 76.0.3809.68 (420c9498db8ce8fcd190a954d51297672c1515d5-refs/branch-heads/3809@{#864})
```

**Comment by mhanberg** [comment]
@AndroidOatmeal Do you mind sharing the contents of some of your failing tests and your support code? I think I found you on the Elixir Slack, so you can share it with me there if you like. Thanks!

**Comment by AdamT** [comment]
I'm not sure if this is related but I ran into this error today:

```
 5) test user can create account (MyAppWeb.Integration.UserRegistrationTest)
     test/my_app_web/integration/user_registration_test.exs:8
     ** (RuntimeError) Wallaby had an internal issue with HTTPoison:
     %HTTPoison.Error{id: nil, reason: :econnrefused}
     %HTTPoison.Error{id: nil, reason: :closed}
     code: tr_fill_in_user_email(page, email)
     stacktrace:
       (wallaby) lib/wallaby/httpclient.ex:37: Wallaby.HTTPClient.make_request/5
       (wallaby) lib/wallaby/experimental/selenium/webdriver_client.ex:52: Wallaby.Experimental.Selenium.WebdriverClient.set_value/2
       (wallaby) lib/wallaby/driver/log_checker.ex:6: Wallaby.Driver.LogChecker.check_logs!/2
       (wallaby) lib/wallaby/element.ex:178: Wallaby.Element.set_value/2
       (wallaby) lib/wallaby/browser.ex:673: Wallaby.Browser.find/3
       test/my_app_web/integration/user_registration_test.exs:12: (test)

21:20:01.785 [error] GenServer Wallaby.SessionStore terminating
** (RuntimeError) Wallaby had an internal issue with HTTPoison:
%HTTPoison.Error{id: nil, reason: :econnrefused}
    (wallaby) lib/wallaby/httpclient.ex:37: Wallaby.HTTPClient.make_request/5
    (elixir) lib/enum.ex:789: anonymous fn/3 in Enum.each/2
    (stdlib) maps.erl:232: :maps.fold_1/3
    (elixir) lib/enum.ex:1964: Enum.each/2
    (stdlib) gen_server.erl:673: :gen_server.try_terminate/3
    (stdlib) gen_server.erl:858: :gen_server.terminate/10
    (stdlib) proc_lib.erl:249: :proc_lib.init_p_do_apply/3
Last message: {:DOWN, #Reference<0.1504753928.1884291075.192258>, :process, #PID<0.554.0>, :shutdown}
```

I ran `mix test` which I never do. I typically run `mix test --trace test/`.

Experimenting a little more I get:
```
# fails most of the time, seems to pass every once in a while
mix test

# passes every time reliably
mix test --trace
```

🤷‍♂

**Comment by jethro-riosa** [comment]
Any updates on this one? I also experiencing this kind of error when running  
`docker run -d -p 4444:4444 -v /dev/shm:/dev/shm selenium/standalone-chrome:3.141.59-zirconium`
locally.

**Comment by aaronrenner** [comment]
@AndroidOatmeal @mhanberg I believe #518 should fix this problem. Would you mind trying again from master?

**Comment by AndroidOatmeal** [comment]
@aaronrenner @mhanberg Thanks so much for putting this together. I updated my wallaby mix dependency to grab the latest version in `master`, which includes PR #518, like so:

```
{:wallaby, github: "elixir-wallaby/wallaby", branch: "master",  only: :test, runtime: false},
```

Unfortunately, I'm still experiencing the same `:econnrefused` error we've been seeing for a while. I also tried the wallaby fix on @mhanberg's [wallaby_stress_test](https://github.com/mhanberg/wallaby_stress_test) repo. Sadly, the issue also persists there. :\

**Comment by aaronrenner** [comment]
I just tried out @mhanberg's wallaby_stress_test and see what you mean. I also found if I switched the driver to phantomjs, there wasn't that problem. The phantom driver uses a pool of phantomjs instances where the chromedriver one uses a single instance. I wonder if the stress test is just overwhelming the underlying chromedriver server and if building a pool of chromedriver instances would help? 🤔

**Comment by mhanberg** [comment]
What is strange is that even if you reduce the number of test modules to 2, it still explodes.

I had the same thought regarding overwhelming chromedriver, but it is strange that chromedriver would crash with only two tests.

**Comment by aarongraham** [comment]
I had a go at building a pool of chromedriver instances as @aaronrenner suggested. I think I may have had some success. I tried it on both my own test suite and on @mhanberg's [wallaby_stress_test](https://github.com/mhanberg/wallaby_stress_test) project and it seems to work. I was able to run tests with a concurrency equal to the number of cpu cores (in my case 20).

I'm not at all familiar with poolboy, or the internals of wallaby, so it would be great if somebody with this experience could review [my fork](https://github.com/aarongraham/wallaby) and make sure I'm going in the right direction. I would also be interested in hearing if this fixes the problem for everybody, or just me.

What I've done so far is rough and untested but my employer is very keen to see this problem fixed and we're happy to invest the time to get a clean PR submitted as soon as possible if people think it would be accepted?

A couple of things I definitely need clarification on:

1. In testing I found that chromedriver can take a very long time to reach a responsive state, sometimes over 5 seconds, by which time the test has failed. To combat this I made the init function in `Chrome.ChromeDriver` block until `ReadinessChecker` confirms ready. I feel like this implementation is preferable anyway as the long startup time happens before any tests begin, and poolboy isn't handing out chromedrivers until they are ready. Am I missing a reason not to do this? 

1. After starting chromedrivers in parallel I found that HTTPoison was running out of connections. At the moment the hackney pool is hardcoded to 4, would there be any reason to not make it equal to the number of chromedriver instances that get created? Or have them both sit under a configuration option?

**Comment by aaronrenner** [comment]
@aarongraham Thanks so much for looking into this! I don't see where having a pool of Chromedriver servers would hurt, and I'm happy to hear it addresses your issue!


> In testing I found that chromedriver can take a very long time to reach a responsive state, sometimes over 5 seconds, by which time the test has failed. To combat this I made the init function in Chrome.ChromeDriver block until ReadinessChecker confirms ready. I feel like this implementation is preferable anyway as the long startup time happens before any tests begin, and poolboy isn't handing out chromedrivers until they are ready. Am I missing a reason not to do this?

Supervisions wait for each child's init to complete before starting the next child, so blocking in the `init/1` callback means the servers would be started sequentially instead of in parallel. (So 20 instances at 5 seconds a piece would take 100 seconds before the tests could start).

To address for phantomJS, I called `Server.wait_until_ready/2` in `Phantom.start_session/1`.

https://github.com/elixir-wallaby/wallaby/blob/53befe04d8445ed0de90770bb20bb884f4586220/lib/wallaby/phantom.ex#L118-L135

This allows everything to start up in parallel and the waiting only occurs when `start_session` is called before the server is ready.

> After starting chromedrivers in parallel I found that HTTPoison was running out of connections. At the moment the hackney pool is hardcoded to 4, would there be any reason to not make it equal to the number of chromedriver instances that get created? Or have them both sit under a configuration option?

That sounds good to me.

Thanks for doing some research on this and I'd be happy to play around with it as well. Please keep us posted on what you find!

**Comment by aaronrenner** [comment]
@aarongraham I just pushed a basic chromedriver pool on this branch that gets wallaby_stress_test passing: https://github.com/elixir-wallaby/wallaby/tree/ar-chromedriver-pool. It definitely slows down things quite a bit, but it seems like with some tuning we might get a speed benefit.

I've also seen where chromedriver leaves chrome processes laying around after it closes. I'm not sure if its a bug in chromedriver, or if its the way our wrapper script is closing chromedriver improperly. However, the end result is it ends up slowing down the computer to a crawl if you run the test suite several times and don't kill the chrome instances afterwards.

**Comment by eahanson** [comment]
@aaronrenner it's possible that there might be some useful code in [Chroxy](https://github.com/holsee/chroxy), which "uses Elixir processes and OTP supervision to manage the chrome instances, as well as including a transparent proxy to facilitate automatic initialisation and termination of the underlying chrome page based on the upstream connection lifetime".

**Comment by AndroidOatmeal** [comment]
@aaronrenner @aarongraham Happy Friday all. :) What's the status of the Chromedriver Pool branch? I've been using `ar-chromedriver-pool` to enable parallel browser tests and it seems to be working. 

It would be nice to merge the fix into the main wallaby branch and finally close this issue. What work remains to be done? I'm happy to help out if the branch needs a bit more polish.

**Comment by filipevarjao** [comment]
I'm getting the same issue. It seems totally random since last week my tests were passing, and this week I'm getting 
```
(RuntimeError) Wallaby had an internal issue with HTTPoison:
     %HTTPoison.Error{id: nil, reason: :econnrefused}
```

**Comment by aswinmohanme** [comment]
Ran into this error today around 5 feature tests. `mix test --max-cases=2` seems to make it work, even though a lot slower than running everything async.

**Comment by padjo** [comment]
had this error 
" test login  --core test (Devato.Features.IndexPageTest)
     test/devato_web/features/index_page_test.exs:13
     ** (RuntimeError) Wallaby had an internal issue with HTTPoison:
     %HTTPoison.Error{id: nil, reason: :timeout}
     stacktrace:
       (wallaby 0.23.0) lib/wallaby/httpclient.ex:37: Wallaby.HTTPClient.make_request/5
       (wallaby 0.23.0) lib/wallaby/experimental/chrome.ex:96: Wallaby.Experimental.Chrome.start_session/1
       (wallaby 0.23.0) lib/wallaby.ex:78: Wallaby.start_session/1
       (devato 0.1.0) test/support/feature_case.ex:25: Devato.FeatureCase.__ex_unit_setup_0/1
       (devato 0.1.0) test/support/feature_case.ex:1: Devato.FeatureCase.__ex_unit__/2
       test/devato_web/features/index_page_test.exs:1: Devato.Features.IndexPageTest.__ex_unit__/2"

i could run the tests chrome headless but i needed it to work so i can see it running on the browser . What worked for me was configuring the timeouts 

https://hexdocs.pm/wallaby/0.23.0/readme.html#config
added
config :wallaby,
  hackney_options: [timeout: :infinity, recv_timeout: :infinity]

# Overriding a value
config :wallaby,
  hackney_options: [timeout: 5_000]

hope this helps others

**Comment by onelesd** [comment]
@mhanberg @aaronrenner @aarongraham 

Sorry to @ you all, but I'm wondering if there is interest in merging `ar-chromedriver-pool`? The branch isn't really usable today because it's deps are so far behind.

Would you accept a PR merging that branch with `main`?

**Comment by AndroidOatmeal** [comment]
@mhanberg @aaronrenner @aarongraham

Hi guys. Sorry to be a pain, but I'd like a bit more visibility into the status of this. This particular issue affects many many users, and has remained open for **four years** now, with no solution or workaround in sight.

If there are technical limitations here, I understand — that may affect our ability to resolve this issue. Can we at least get an update explaining the current status of this particular issue? Much appreciated 🙏

**Comment by mhanberg** [comment]
I'll get re acquainted with this and take another look.

**Comment by AndroidOatmeal** [comment]
@mhanberg Thank you, I appreciate it. I can readily reproduce this error when running acceptance tests in parallel. I'm happy to assist if needed.

**Comment by mhanberg** [comment]
Alright everyone, I think this can finally be closed.

#692 should be the final fix and has been released as part of [v0.30.0](https://github.com/elixir-wallaby/wallaby/releases/tag/v0.30.0).

Please note that the minimum Elixir version for Wallaby has been increased to v1.12. If this is a serious problem for you, please open an issue and we can discuss.

If you think that you are still seeing this problem, please open another issue. This one is too long and the codebase has shifted so much since it was created that a lot of the code examples and errors shared here are out of date/no longer relevant.

**Comment by kadimi** [comment]
I'm on Ubuntu... This is what fixed it for me:

```sh
sudo snap alias chromium.chromedriver chromedriver
```
